File: F# Workspace/01_Basics/01_Basics/factorial_vers.fs
(*
 * factorial_vers.fs
 * By Antonio F. Huertas
 * This program calculates and displays the factorial of the
 * given non negative integer.
 *)

// Uses an if experssion.
let rec factorial num =
    if num = 0
        then 1
        else num * factorial (num - 1)

// Uses a pattern matching expression.
let rec factorial_v2 num =
    match num with
    | 0 -> 1
    | _ -> num * factorial_v2 (num - 1)

// Uses a pattern matching function.
let rec factorial_v3 = function
    | 0 -> 1
    | num -> num * factorial_v3 (num - 1)

// Uses a tail recursive helper function.
let factorial_v4 num =
    let rec helper acc num =
        match num with
        | 0 -> acc
        | _ -> helper (acc * num) (num - 1)
    helper 1 num

// start : int -> unit
// Serves as an entry point for the program.
let start num =
    printfn "The number is %d" num
    if num < 0 then
        printfn "Error! Negative integer."
    else
        printfn "Using a tail-recursive helper, the factorial is %d" 
            (factorial_v4 num)
        printfn "Using a pattern matching function, the factorial is %d"
            (factorial_v3 num)
        printfn "Using a pattern matching expression, the factorial is %d" 
            (factorial_v2 num)
        printfn "Using an if expression, the factorial is %d"
            (factorial num)

File: Prolog Workspace/01_Basics/factorial_vers.pl
/*
 * factorial_vers.pl
 * By Antonio F. Huertas
 * This program displays the factorial of a non-negative integer
 * using different versions.
 */

%% factorial(+Number, ?Fact)
%  Computes the factorial of a non-negative integer.
%  NOTE: This version uses regular recursion.
factorial(0, 1).

factorial(Number, Fact) :-
                  Number > 0,
                  Pred is Number - 1,
                  factorial(Pred, FactPred),
                  Fact is Number * FactPred.

%% factorial_v2(+Number, ?Fact)
%  Computes the factorial of a non-negative integer.
%  NOTE: This version uses a helper tail-recursive predicate.
factorial_v2(Number, Fact) :-
                     factorial_v2_helper(Number, 1, Fact).

% factorial_v2_helper(+Number, +Accum, ?Fact)
factorial_v2_helper(Number, Accum, Fact) :-
                            Number > 0,
                            NewAccum is Accum * Number,
                            NewNumber is Number - 1,
                            factorial_v2_helper(NewNumber, NewAccum, Fact).
                            
factorial_v2_helper(0, Accum, Accum).

%% start(+Number)
%  Serves as an entry point for the program.
start(Number) :-
    write('The number is '), write(Number), nl,
    Number >= 0,

    factorial_v2(Number, Fact2),
    write('Tail recursion: the factorial is '), write(Fact2), nl,

    factorial(Number, Fact),
    write('Regular recursion: the factorial is '), write(Fact), nl.
